app/ota_client.py

```diff
@@ -896,48 +892,138 @@ def _gen_regular_files(self, rootfs_dir, regulars_file, target_dir):
+            return False
+        finally:
+            os.chdir(cwd)
+
+        return True
+
+    def _process_regular_files_pool_init(self, gvar_dict):
+        ''' 
+            Used by _process_regular_files
+            Init the worker pool with shared variables
+
+            DO NOT call this method from other functions 
+            except for _process_regular_files
+        '''
+        global global_var_dict
+        global_var_dict = gvar_dict
+
+    def _process_regular_files_exit(self, gvar_dict):
+        '''
+            Used by _process_regular_files
+            Update corresponding class attributes
+
+            DO NOT call this method from other functions 
+            except for _process_regular_files
+        '''
+        for k, v in gvar_dict:
+            # apply staged data to target attribute
+            vname = k.split("-")
+            if vname[0] == "staging":
+                if vname[1] == "dict":
+                    # TODO: update or replace?
+                    setattr(self, vname[-1], dict(v))
+                elif vname[1] == "list":
+                    setattr(self, vname[-1], list(v))
+
+    def _process_regular_files_log2file_listener(self, flog_path, log_queue, event):
+        '''
+            logger worker for _process_regular_files
+        '''
+        with open(flog_path, "w") as f:
+            while not event.is_set() or not log_queue.empty():
+                    msg = log_queue.get()
+                    f.write(msg+"\n")
+
+    def _process_regular_files(self, rootfs_dir, rfiles_list, target_dir):
+        with Manager() as manager, \
+            tempfile.NamedTemporaryFile(delete=False) as flog:
+            log_file_path, log_q, ev = flog.name, manager.Queue(), manager.Event()
+            # variables passed to child processes
+            # variable naming pattern: <prefix>-<type>-<var_name>
+            # prefix tmp: for temporary use
+            # prefix staging: used to update corresponding class attribute
+            gvar_dict = {
+                "tmp-queue-log_queue": log_q,
+                "tmp-dict-hardlink_reg": manager.dict(), 
+                "staging-dict-_rollback_dict": manager.list()
+            } 
+        
+            # launch log2file logger
+            #   passing file descriptor to sub-process is not possible,
+            #   so we pass the log file path to the logger, and let the logger
+            #   open the log file by itself
+            p_logger = Process(
+                target=self._process_regular_files_log2file_listener, 
+                args=(log_file_path, log_q, ev, )
+                )
+            p_logger.start()
+
+            # default to one worker per CPU core
+            with Pool(
+                initializer=self._process_regular_files_pool_init,
+                initargs=(gvar_dict,)
+            ) as pool:
+                for rfile_inf in rfiles_list:
+                    pool.apply_async(
```

Comment:
return value of `pool.apply_async` should be checked by get?
like:
```suggestion
                    results = [pool.apply_async(...) for rfile_inf in rfiles_list]
                    for result in results:
                        result.get()
```

Reply:
In commit a20f92c, I update the logic of how sub processes notifying the main process about the exception,
The worker process needs not to return anything.

If everything is OK, the worker process(**_process_regular_file**) will terminate silently without triggering anything.
If an exception is triggered in a sub process, an error_callback function will be triggered to catch that exception, and then pass it to the main process(**_process_regular_files**). The main process will raise a new exception to notify the upper caller that the regular files processing failed and terminated.

To ensure the whole OTA operation is atomic, all sub processes will also be terminated by any exception being triggered.